README
$Id: README 15178 2015-11-06 10:45:03Z fwobbe $

== Information for installing GMT 5 ==

Note: The build system has recently been switched to CMake which is a
cross-platform system for managing the build process. If you are familiar
with the old GNU Build Tools (automake, autoconf, and configure) you can
probably skip over to the CMake quick start guide in README.CMake.


Introduction
------------

You do not need to read these instructions unless you plan to build and
install the programs manually.

GMT has been installed successfully under UNIX/Linux/OS X on workstations.  It
also installs under Windows and in UNIX emulators such as Cygwin or on virtual
machines.  We anticipate few problems if you are installing the package on
other platforms.

Note there are three GMT tar archives available (#3 is optional):

1. gmt-5.x.x.tar.bz2:          The GMT 5 distribution
2. gshhg-gmt-2.x.x.tar.gz:     All five resolutions of GSHHG coastline data
3. dcw-gmt-1.x.x.tar.bz2:      Digital Chart of the World polygon data

The archives are available in bzip2 (*.bz2) and gzip (*.gz) formats.
If you do not have bzip2 installed you can obtain source or executables
from http://www.bzip.org.

For Windows users there are separate Windows installers available; this
discussion only considers UNIX/Linux/OS X installations. Windows users who
which to build GMT from the sources refer to README.WIN32.


CMake
-----

Install CMake (>=2.8.5) from http://www.cmake.org/cmake/resources/software.html


Install netCDF library
----------------------

For all major Linux distributions there are prepackaged development binaries
available. netCDF is also available on MacOSX trough the macports and fink
package managers.

Otherwise, get netCDF from http://www.unidata.ucar.edu/downloads/netcdf/.
You need at least version 4.0 with netCDF-4/HDF5 data model support (do not
disable HDF5/ZLIB in netCDF with --disable-netcdf-4).


GDAL (optional)
---------------

To use the GDAL interface (ability to provide grids or images to be imported
via gdal) you must have the GDAL library and include files installed.  Like
netCDF, GDAL is available through your favorite package manager on many *NIX
systems.


PCRE (optional)
---------------

To use the PCRE interface (ability to specify regular expressions in some
search options, e.g., gmtconvert) you must have the PCRE library and include
files installed.  PCRE is available through your favorite package manager
on many *NIX systems.


LAPACK (optional)
---------------

To greatly speed up some linear algebra calculations (greenspline in 
particular) you must have the LAPACK library and include files installed.
LAPACK is available through your favorite package manager on many *NIX
systems or in the case of OS X is built in.


Install support data
--------------------

You can obtain GMT from http://gmt.soest.hawaii.edu/. Alternatively you may
get GMT from any of the following FTP sites. Try the site that is closest to
you to minimize transmission times:

Site                                    FTP address
--------------------------------------------------------------
SOEST, U. of Hawaii                     ftp.soest.hawaii.edu
Lab for Satellite Altimetry, NOAA       ibis.grdl.noaa.gov
IRIS, Washington, US                    ftp.iris.washington.edu
IAG-USP, U. of Sao Paulo, BRAZIL        ftp.iag.usp.br
Dep. Geology, U. in Oslo, NORWAY        ftp.geologi.uio.no
ISV, Hokkaido U, Sapporo, JAPAN         ftp.eos.hokudai.ac.jp
GDS, Vienna U. of Technology, AUSTRIA   gd.tuwien.ac.at
TENET, Tertiary Education & Research Networks, SOUTH AFRICA
                                        gmt.mirror.ac.za

The development sources are available from the subversion repository at
svn://gmtserver.soest.hawaii.edu/gmt5/trunk

Extract the files and put them in a separate directory (need not be
where you eventually want to install GMT).


Configuring
-----------

GMT can be build on any platform supported by CMake.  CMake is a
cross-platform, open-source system for managing the build process.
Refer to README.CMake for further details.  In the source tree copy
cmake/ConfigUserTemplate.cmake to cmake/ConfigUser.cmake and edit
the file according to your demands.

By default, GMT will use Dave Watson's Delaunay triangulation routine.
However, a much faster alternative is available from Jonathan Shewchuk, but
his routine is not distributed under the GNU Public License.  If you work for
a for-profit organization you should read Shewchuk's copyright statement (in
src/triangle.c) first.  If you agree with the license terms you can enable
Shewchuk's triangulation routine in cmake/ConfigUser.cmake.

At run-time, GMT will initialize all default variables by reading the
share/gmt.conf. Although strongly discouraged, if you want to make permanent
changes to some of the defaults you can edit the share/gmt.conf file directly.
A much better approach is to create a gmt.conf file in your home directory
and edit those settings since GMT will check for that file before loading
system defaults (actually, it will first look in the current directory, then
the home directory, and then finally in share).  See the gmt.conf man page
for a description of all defaults.

To prevent two GMT processes writing to the same gmt.conf file simultaneously
(thereby corrupting it), GMT can implement the POSIX advisory file locking
scheme and sets and releases locks on these files.  This might not be reliable
when the files reside in directories on network filesystems, such as NFS.
Whether flock works on network filesystems is implementation dependent.  If
you want to activate file locking you may enable it in cmake/ConfigUser.cmake.

By default, both GMT and all its supplements are built.  You can turn
off all supplements via the BUILD_SUPPLEMENTS setting in ConfigUsers.cmake

The top-level installation directory is configured with the variable
CMAKE_INSTALL_PREFIX.

Now that you made your configuration choices it is time for invoking CMake.
Create a subdirectory where the build files will be generated, e.g., in the
source tree 'mkdir build'.

In the build subdirectory, type

  cmake [options] ..

Append any of the options explained above as you see fit.  If CMake cannot
figure out all the dependent libraries or required compiler and linker flags
it will give you a message and you will be asked to edit
cmake/ConfigUser.cmake.


Build GMT
---------

In the build directory, type

  make

which will compile all the programs.  After a successful compilation you may
install the executables in the designated bin directory with the command

  make install

After a successful install you can have the object files and the local
executables removed by saying

  make clean

or just remove the entire build directory.


Documentation
_____________

The documentation is available online at http://gmt.soest.hawaii.edu/
or as platform independent package that you can install along with GMT.

The GMT documentation includes HTML files for online browsing, user guide,
cookbook, and manual pages. The Documentation also contains the
GMT_Tutorial.pdf file which is a short course in how to use GMT.  It can be
^^^^^^^^^^^^^^^^
still missing!
followed individually or in a lab setting by a group of users.

The development sources from subversion do not contain the precompiled
documentation. The manuals, HTML pages, and PDFs have to be created from
source with Sphinx (see README.CMake).


Set path
--------

Make sure users set their PATH to include the directory containing
the GMT executables (BINDIR) if this is not a standard directory
like /usr/local/bin.  You should now be able to run GMT programs.


GMT supplemental Code
---------------------

GMT users elsewhere have developed programs that utilize the GMT libraries and
produce PostScript code compatible with the rest of GMT or simply perform data
manipulation.  In addition, misc.  code developed by us depend on GMT
libraries.  Currently, the supplemental archive include the directories:

  gshhg     - Data extractor for GSHHG shoreline polygons and rivers, borders.
  img       - Data extractor for Smith/Sandwell altimetry grids.
  meca      - Plotting of focal mechanisms, velocity arrows,
              and error ellipses on maps.
  mgd77     - Programs for handling of native MGD77 files.
  misc      - dimfilter
  potential - geopotential manipulations
  segyprogs - Plotting SEGY seismic data sets.
  spotter   - Plate tectonic & kinematics applications.
  x2sys     - Track intersection (crossover) tools.


Misc
----

Before running programs, there are a few things you should do/know:

  - Read carefully the documentation for the gmt system.  This can be
    found as both PDF and HTML files in the doc/pdf|html directories.
    The successful operation of gmt-programs depends directly on your
    understanding of how gmt "works", its option lists, I/O, and composite
    plot mechanisms.  Then, before running individual gmt programs, read
    the associated man page.


A reminder
----------

If you think it is appropriate, you may consider paying us back by including
our latest EOS article (Wessel, P., W. H. F. Smith, R. Scharroo, J. F. Luis,
and F. Wobbe (2013), Generic Mapping Tools: Improved version released, Eos
Trans. AGU, 94(45), 409-410, doi:10.1002/2013EO450001) in the reference list
of your future publications that will benefit from the availability of GMT.


Software support
----------------

You haven't bought anything so you cannot expect full service.  However, if
you find a bug in any of the programs, please report it to us
(http://gmt.soest.hawaii.edu/) rather than trying to fix it yourself so that
we, and through us, other users may benefit from your find.  Make sure you
provide us with enough information so that we can recreate the problem.

In addition to the bug tracking feature (New Issues) on the website, you
can also post general questions on the GMT user forum.  Note that registration
is required to post on the site.


Ordering the GMT package on CD/DVD-Rs
-------------------------------------

Should you or someone you know without net-access need to obtain GMT:
Geoware makes and distributes CD/DVD-Rs with the GMT package and many
useful data sets.  For more details and a full description of the data
sets (up to 60 Gb of data!) visit http://www.geoware-online.com/.


Good luck!

The GMT Team.

vim: tw=78 et ts=2 sts=2 sw=2 autoindent smartindent


ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 15178 2015-11-06 10:45:03Z fwobbe $
-----------------------------------------------------------------------------------------

2015-01-08	pwessel
	* gmtregress.c		New tool for a variety of 1-D line regression.

2014-12-03	pwessel
	* grdconvert.c		The new name of grdreformat, backwards compatible.

2014-09-08	pwessel
	* ps2raster.c:		For non-PDF output, only handled transparency via @<transp>
				and not via -t<transp> option.

2014-08-10	pwessel
	* gmt_init.c:		Allow UTM projections to not set the zone; if so then the
				zone is automatically determined from the region
	* gmt_map.c:		Determine best UTM zone from region if no zone is set.
	* gmt_parse.c:		Allow new -R flavor for grid creation: -R[L|C|R][T|M|B]<x0>/<y0>/<nx>/<ny>,
				where one specifies a fixed point relative to a rectangular
				domain using the justification codes, and then adds the
				dimensions of the new grid.  The grid spacings must be
				set via the -I option.
	* psxy[z].c:		Added optional arg to -Sv|m +b+e modifiers to set the kind
				of arrow head to use.  Choose from t (termimal line),
				c (circle) or a (arrow) [Default].
	* grdvector.c:		---"---
	
2014-07-17	pwessel
	* psxy[z].c:		Added modifier +p[<pen>] to the front specification so that
				one can specify an alternate symbol outline, or no outline
				at all (by not specifying <pen>).

2014-06-07	pwessel
	* ps2raster.c:		Added new device -Ts for SVG (scalable vector graphics).

2014-05-30	pwessel
	* grdtrack.c:		Did not properly handle the distance units specified in -E.
				Now we pre-scan for units to ensure that only ONE unit is used
				for the various distance options, then apply that unit in the
				distance calculations throughout.  Results will use that unit.
				If no unit is specified we default to km (geographic) or user
				unit (Cartesian).  Also required some changes in gmt_support.c
				Response to forum message # 689.

2014-05-29	pwessel
	* gmt_plot.c:		The length of the map scale bar was incorrectly calculated
				for some oblique projections.  Issue #569.

2014-05-26	pwessel
	* gmt_map.c:		The map region clip path for -JE/e was hardwired to a full
				circle and did not care about -R setting.
				When -R...r and -Je?E was given the search for enclosing
				geographic boundaries failed when either the S or N pole
				was the projection's antipole.
	* gmt_plot.c, pslib.?, pstext.c, psclip.c:	New handling of PSL clipping for quoted text
				lines (contours) and text.  No more different clip off types
				in psclip.

2014-05-20	remko
	* *.[ch]:		Renamed TIME_LANGUAGE to GMT_LANGUAGE. Introduced language-
				dependent names and abbreviations of cardinal points.

2014-05-08	remko
	* gmt_plot.c:		Fix bug in treatment of variable text introduced on 2014-05-04.
				This addresses issue #554.

2014-05-07	pwessel, remko
	* gmt_support.c:	Added support for both variables and constants on both sides
				of the operator in the conditional tests used for custom symbols;
				see issue # 554.
	* gmt_plot.c:		---"---
	* gmt_symbol.h:		___"___

2014-05-04	pwessel
	* gmt_plot.c:		Fix treatment of text and font attributes in custom symbols
	* gmt_support.c:	---"---.  This addresses issue # 553.

2014-04-24	pwessel
	* grdcut.c:		Needed to extend -Z to also consider -Zr so that -Zn
				would return a NaN-free subset.

2014-04-05	pwessel
	* xyz2grd.c:		Replaced -N<nodata> with common option -di<nodata>, but
				left -N for backwards compatibility under 4.
	* grd2xyz.c:		Replaced -N[i]<nodata> with common option -d[i]<nodata>, but
				left -N for backwards compatibility under 4.

2014-04-04	pwessel
	* psbasemap.c:		Added new option -A[file] which instead of plotting a map
				will write the coordinates of the plot domain outline to
				the given file (or stdout).

2014-04-03	pwessel
	* gmt_init.c:		Added new common GMT option -d[i|o]<nodata>, which can be
				used anywhere to convert users <nodata> values to NaN, and to do
				the reverse on output.

2014-04-02	pwessel
	* spotter/backtracker.c:	Merge -e with -E to simplify operations and to
				disallow lower-case options other than for GMT common options.
				Backwards compatible for now but this will cease in 5.2.x
	* spotter/grdrotator.c:	---"---

2014-03-26	remko
	* ps2raster.c:		Bug fix: -C should also apply when computing BoundingBox.

2014-03-03	pwessel
	* pslegend.c:		The flush_paragraph was not reset to false before 2nd loop,
				causing the text (T) option to fail if it was the first item.

-----------------------------------------------------------------------------------------
2014-03-01	pwessel
	GMT 5.1.1 finalized and released.
-----------------------------------------------------------------------------------------
2014-02-27	remko
	* ps2raster.c:		Fix issue with psimage-included EPS files (issue #521)

2014-02-25	pwessel
	* xyz2grd.c:		Gave incorrect n_read count and for -bi -An insisted on 3 cols.

2014-02-22	pwessel
	* grdcontour.c:		Fix writing of contour labels to text file.

2014-02-20	pwessel
	* grdcontour.c:		The -Zscale/offset option was applied incorrectly due to a
				change in the underlying scaling function.
	* grdvolume.c:		---"---

2014-02-19	pwessel
	*mgd77/mgd77list.c:	Crashed with -Am4 as time column was not loaded.

2014-02-18	pwessel
	*gmtmath.c:		Crashed for some options because a dataset's n_records was not set.

2014-02-10	pwessel
	*gmt2kml.c:		Make sure the default style IDs are unique for each process.

2014-02-06	pwessel
	* psxy.c:		New modifier -Ex[+]y[+] to do asymmetrical error bars.

2014-02-03	pwessel
	* gmtconvert.c:		New option -C to limit output to segments whose record count
				matches our criteria.
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then

src/gmt.c
/*--------------------------------------------------------------------
 *	$Id: gmt.c 15178 2015-11-06 10:45:03Z fwobbe $
 *
 *	Copyright (c) 1991-2015 by P. Wessel, W. H. F. Smith, R. Scharroo, J. Luis and F. Wobbe
 *	See LICENSE.TXT file for copying and redistribution conditions.
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU Lesser General Public License as published by
 *	the Free Software Foundation; version 3 or any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU Lesser General Public License for more details.
 *
 *	Contact info: gmt.soest.hawaii.edu
 *--------------------------------------------------------------------*/
/*
 * Launcher for any GMT5 module via the corresponding function.
 * Modules are loaded dynamically from the GMT core library, the
 * optional supplemental library, and any number of custom libraries
 * listed via GMT_CUSTOM_LIBS in gmt.conf.  If <module> is not found
 * we also try gmt<module> in case user left that part off.
 *
 * Version:	5
 * Created:	17-June-2013
 *
 */

#include "gmt_dev.h"

#if !(defined(WIN32) || defined(NO_SIGHANDLER))
#	include <signal.h>
#	include "common_sighandler.h"
#endif

#define PROGRAM_NAME	"gmt"

/* Determine the system environmetal parameter that leads to shared libraries */
#if defined _WIN32
#define LIB_PATH "%%PATH%%"
#elif defined __APPLE__
#define LIB_PATH "DYLD_LIBRARY_PATH"
#else
#define LIB_PATH "LD_LIBRARY_PATH"
#endif

int main (int argc, char *argv[]) {
	int status = GMT_NOT_A_VALID_MODULE;	/* Default status code */
	int k, v_mode = GMT_MSG_COMPAT;		/* Default verbosity */
	bool gmt_main = false;			/* Set to true if no module specified */
	unsigned int modulename_arg_n = 0;	/* Argument number in argv[] that contains module name */
	unsigned int mode = GMT_SESSION_NORMAL;	/* Default API mode */
	struct GMTAPI_CTRL *api_ctrl = NULL;	/* GMT API control structure */
	char gmt_module[GMT_LEN32] = "gmt";	/* Alternate module name that starts with "gmt" */
	char *progname = NULL;			/* Last component from the pathname */
	char *module = NULL;			/* Module name */

#if !(defined(WIN32) || defined(NO_SIGHANDLER))
	/* Install signal handler */
	struct sigaction act;
	sigemptyset(&act.sa_mask); /* Empty mask of signals to be blocked during execution of the signal handler */
	act.sa_flags = SA_SIGINFO | SA_NODEFER; /* Do not prevent the signal from being received from within its own signal handler. */
	act.sa_sigaction = sig_handler;
	sigaction (SIGINT,  &act, NULL);
	act.sa_flags = SA_SIGINFO;
	sigaction (SIGILL,  &act, NULL);
	sigaction (SIGFPE,  &act, NULL);
	sigaction (SIGBUS,  &act, NULL);
	sigaction (SIGSEGV, &act, NULL);
#endif /* !(defined(WIN32) || defined(NO_SIGHANDLER)) */

	/* Look for and process any -V[flag] so we may use GMT_Report_Error early on.
	 * Because first 2 bits of mode is used for other things we must left-shift by 2 */
	for (k = 1; k < argc; k++) if (!strncmp (argv[k], "-V", 2U)) v_mode = GMT_get_V (argv[k][2]);
	if (v_mode) mode = ((unsigned int)v_mode) << 2;	/* Left-shift the mode by 2 */
	/* Initialize new GMT session */
	if ((api_ctrl = GMT_Create_Session (argv[0], GMT_PAD_DEFAULT, mode, NULL)) == NULL)
		return EXIT_FAILURE;
	api_ctrl->internal = true;	/* This is a proper GMT internal session (external programs will default to false) */
	progname = strdup (basename (argv[0])); /* Last component from the pathname */
	/* Remove any filename extensions added for example
	 * by the MSYS shell when executing gmt via symlinks */
	GMT_chop_ext (progname);

	/* Test if argv[0] contains a module name: */
	module = progname;	/* Try this module name unless it equals PROGRAM_NAME in which case we just enter the test if argc > 1 */
	gmt_main = !strcmp (module, PROGRAM_NAME);	/* true if running the main program, false otherwise */
	if (gmt_main && argc > 1 && (!strcmp (argv[1], "gmtread") || !strcmp (argv[1], "read") || !strcmp (argv[1], "gmtwrite") || !strcmp (argv[1], "write"))) {
		/* Cannot call [gmt]read or [gmt]write module from the command-line */
		module = argv[1];	/* Name of module that does not exist */
		status = GMT_NOT_A_VALID_MODULE;
		modulename_arg_n = 1;
		goto no_such;
	}

	if ((gmt_main || (status = GMT_Call_Module (api_ctrl, module, GMT_MODULE_EXIST, NULL)) == GMT_NOT_A_VALID_MODULE) && argc > 1) {
		/* argv[0] does not contain a valid module name, and
		 * argv[1] either holds the name of the module or an option: */
		modulename_arg_n = 1;
		module = argv[1];	/* Try this module name */
		if ((status = GMT_Call_Module (api_ctrl, module, GMT_MODULE_EXIST, NULL) == GMT_NOT_A_VALID_MODULE) != 0) {
			/* argv[1] does not contain a valid module name; try prepending gmt: */
			strncat (gmt_module, argv[1], GMT_LEN32-4U);
			status = GMT_Call_Module (api_ctrl, gmt_module, GMT_MODULE_EXIST, NULL); /* either GMT_NOERROR or GMT_NOT_A_VALID_MODULE */
			if (status != GMT_NOT_A_VALID_MODULE) module = gmt_module;
		}
	}

	if (status == GMT_NOT_A_VALID_MODULE) {
		/* neither argv[0] nor argv[1] contain a valid module name */
		int arg_n;
		status = GMT_OK; /* default exit status */

		if (argv[1+modulename_arg_n] && !strcmp (argv[1+modulename_arg_n], "=") && argv[2+modulename_arg_n] == NULL) {
			/* Just wanted to know if module exists */
			status = 1; /* Return nonzero when modules does not exist */
			goto exit;
		}

		for (arg_n = 1; arg_n < argc; ++arg_n) {
			/* Try all remaining arguments: */

			/* Print module list */
			if (!strcmp (argv[arg_n], "--help")) {
				fprintf (stderr, "\n\tGMT - The Generic Mapping Tools, Version %s [%u cores]\n", GMT_VERSION, api_ctrl->n_cores);
				fprintf (stderr, "(c) 1991-%d Paul Wessel, Walter H. F. Smith, R. Scharroo, J. Luis, and F. Wobbe\n\n", GMT_VERSION_YEAR);
				fprintf (stderr, "Supported in part by the US National Science Foundation (http://www.nsf.gov/)\n");
				fprintf (stderr, "and volunteers from around the world (see http://gmt.soest.hawaii.edu/).\n\n");

				GMT_Call_Module (api_ctrl, NULL, GMT_MODULE_PURPOSE, NULL);
				goto exit;
			}

			/* Print all modules and exit */
			if (!strcmp (argv[arg_n], "--show-modules")) {
				GMT_Call_Module (api_ctrl, NULL, GMT_MODULE_LIST, NULL);
				goto exit;
			}

			/* Print version and exit */
			if (!strcmp (argv[arg_n], "--version")) {
				fprintf (stdout, "%s\n", GMT_PACKAGE_VERSION_WITH_SVN_REVISION);
				goto exit;
			}

			/* Show number of cores */
			if (!strcmp (argv[arg_n], "--show-cores")) {
				fprintf (stdout, "%u\n", api_ctrl->n_cores);
				goto exit;
			}

			/* Show share directory */
			if (!strcmp (argv[arg_n], "--show-datadir")) {
				fprintf (stdout, "%s\n", api_ctrl->GMT->session.DATADIR);
				goto exit;
			}

			/* Show the directory that contains the 'gmt' executable */
			if (!strcmp (argv[arg_n], "--show-bindir")) {
				fprintf (stdout, "%s\n", api_ctrl->GMT->init.runtime_bindir);
				goto exit;
			}

			/* Show the directory that contains the shared plugins */
			if (!strcmp (argv[arg_n], "--show-plugindir")) {
				fprintf (stdout, "%s\n", api_ctrl->GMT->init.runtime_plugindir);
				goto exit;
			}
			
			/* Show share directory */
			if (!strcmp (argv[arg_n], "--show-sharedir")) {
				fprintf (stdout, "%s\n", api_ctrl->GMT->session.SHAREDIR);
				goto exit;
			}

		} /* for (arg_n = 1; arg_n < argc; ++arg_n) */

		/* If we get here, we were called without a recognized modulename or option
		 *
		 * gmt.c is itself not a module and hence can use fprintf (stderr, ...). Any API needing a
		 * gmt-like application will write one separately [see mex API] */
no_such:

		fprintf (stderr, "\n\tGMT - The Generic Mapping Tools, Version %s [%u cores]\n", GMT_VERSION, api_ctrl->n_cores);
		fprintf (stderr, "(c) 1991-%d Paul Wessel, Walter H. F. Smith, R. Scharroo, J. Luis, and F. Wobbe\n\n", GMT_VERSION_YEAR);
		fprintf (stderr, "Supported in part by the US National Science Foundation (http://www.nsf.gov/)\n");
		fprintf (stderr, "and volunteers from around the world (see http://gmt.soest.hawaii.edu/).\n\n");

		fprintf (stderr, "This program comes with NO WARRANTY, to the extent permitted by law.\n");
		fprintf (stderr, "You may redistribute copies of this program under the terms of the\n");
		fprintf (stderr, "GNU Lesser General Public License (http://www.gnu.org/licenses/lgpl.html).\n");
		fprintf (stderr, "For more information about legal matters, see the file named LICENSE.TXT.\n\n");
		fprintf (stderr, "usage: %s [options]\n", PROGRAM_NAME);
		fprintf (stderr, "       %s <module name> [<module-options>]\n\n", PROGRAM_NAME);
		fprintf (stderr, "options:\n");
		fprintf (stderr, "  --help            List a description of available GMT modules.\n");
		fprintf (stderr, "  --show-bindir     Show directory with GMT executables.\n");
		fprintf (stderr, "  --show-cores      Show number of available cores.\n");
		fprintf (stderr, "  --show-datadir    Show directory/ies with user data.\n");
		fprintf (stderr, "  --show-modules    Single list of all module names.\n");
		fprintf (stderr, "  --show-plugindir  Show directory for plug-ins.\n");
		fprintf (stderr, "  --show-sharedir   Show directory for shared GMT resources.\n");
		fprintf (stderr, "  --version         Print GMT version number.\n\n");
		fprintf (stderr, "if <module-options> is \'=\' we call exit (0) if module exist and non-zero otherwise.\n\n");
		if (modulename_arg_n == 1 && module[0] != '-') {
			fprintf (stderr, "ERROR: No module named %s was found.  This could mean one of three things:\n", module);
			fprintf (stderr, "  1. There actually is no such module; check your spelling.\n");
			if (strlen (GMT_SUPPL_LIB_NAME))
				fprintf (stderr, "  2. Module exists in the GMT supplemental library, but the library could not be found.\n");
			else
				fprintf (stderr, "  2. Module exists in the GMT supplemental library, but the library was not installed.\n");
			if (api_ctrl->n_shared_libs > 2)
				fprintf (stderr, "  3. Module exists in a GMT custom library, but the library could not be found.\n");
			else
				fprintf (stderr, "  3. Module exists in a GMT custom library, but none was specified via GMT_CUSTOM_LIBS.\n");
			fprintf (stderr, "Shared libraries must be in standard system paths or set via environmental parameter %s.\n\n", LIB_PATH);
		}
		status = EXIT_FAILURE;
		goto exit;
	} /* status == GMT_NOT_A_VALID_MODULE */

	/* Here we have found a recognized GMT module and the API has been initialized. */
	if (argv[1+modulename_arg_n] && !strcmp (argv[1+modulename_arg_n], "=") && argv[2+modulename_arg_n] == NULL)	/* Just want to know if module exists */
		status = GMT_OK;
	else {	/* Now run the specified GMT module: */
		if ((argc-1-modulename_arg_n) == 0)	/* No args, call explicitly with NULL because under Cygwin argv[2] may not be NULL */
			status = GMT_Call_Module (api_ctrl, module, 0, NULL);
		else
			status = GMT_Call_Module (api_ctrl, module, argc-1-modulename_arg_n, argv+1+modulename_arg_n);
	}

exit:
	if (progname) free (progname);
	/* Destroy GMT session */
	if (GMT_Destroy_Session (api_ctrl))
		return EXIT_FAILURE;

	return status; /* Return the status from the module */
}
